include/zephyr/drivers/clock_management/clock_driver.h

```diff
@@ -0,0 +1,410 @@
+ * @ref clock_notify_children, and if the return code of that call is
+ * ``CLK_NO_CHILDREN`` the clock may safely power itself down.
+ * @param clk_hw Clock object to notify of reconfiguration
+ * @param parent Parent clock device that was reconfigured
+ * @param event Clock reconfiguration event
+ * @return -ENOSYS if clock does not implement notify_children API
+ * @return -ENOTSUP if clock child cannot support new rate
+ * @return -ENOTCONN to indicate that clock is not using this parent. This can
+ *         be useful to multiplexers to indicate to parents that they may safely
+ *         shutdown
+ * @return negative errno for other error notifying clock
+ * @return 0 on success
+ */
+static inline int clock_notify(const struct clk *clk_hw,
+			       const struct clk *parent,
+			       const struct clock_management_event *event)
+{
+#ifdef CONFIG_CLOCK_MANAGEMENT_RUNTIME
+	if (!(clk_hw->api) || !(clk_hw->api->notify)) {
+		return -ENOSYS;
+	}
+
+	return clk_hw->api->notify(clk_hw, parent, event);
+#else
+	return -ENOTSUP;
+#endif
+}
+
+/**
+ * @brief Configure a clock
+ *
+ * Configure a clock device using hardware specific data. This must also
+ * trigger a reconfiguration notification for any consumers of the clock.
+ * Called by the clock management subsystem, not intended to be used directly
+ * by clock drivers
+ * @param clk_hw clock device to configure
+ * @param data hardware specific clock configuration data
+ * @return -ENOSYS if clock does not implement configure API
+ * @return -EIO if clock could not be configured
+ * @return -EBUSY if clock cannot be modified at this time
+ * @return negative errno for other error configuring clock
+ * @return 0 on successful clock configuration
+ */
+static inline int clock_configure(const struct clk *clk_hw, const void *data)
+{
+	int ret;
+
+	if (!(clk_hw->api) || !(clk_hw->api->configure)) {
+		return -ENOSYS;
+	}
+
+	ret = clk_hw->api->configure(clk_hw, data);
+#ifdef CONFIG_CLOCK_MANAGEMENT_CLK_NAME
+	LOG_MODULE_DECLARE(clock_management, CONFIG_CLOCK_MANAGEMENT_LOG_LEVEL);
+	LOG_DBG("Clock %s reconfigured with result %d", clk_hw->clk_name, ret);
+#endif
+	return ret;
+}
+
+/**
+ * @brief Get rate of a clock
+ *
+ * Gets the rate of a clock, in Hz. A rate of zero indicates the clock is
+ * active or powered down.
+ * @param clk_hw clock device to read rate from
+ * @return -ENOSYS if clock does not implement get_rate API
+ * @return -EIO if clock could not be read
+ * @return negative errno for other error reading clock rate
+ * @return frequency of clock output in HZ
+ */
+static inline int clock_get_rate(const struct clk *clk_hw)
+{
+	int ret;
+
+	if (!(clk_hw->api) || !(clk_hw->api->get_rate)) {
+		return -ENOSYS;
+	}
+
+	ret = clk_hw->api->get_rate(clk_hw);
+#ifdef CONFIG_CLOCK_MANAGEMENT_CLK_NAME
+	LOG_MODULE_DECLARE(clock_management, CONFIG_CLOCK_MANAGEMENT_LOG_LEVEL);
+	LOG_DBG("Clock %s returns rate %d", clk_hw->clk_name, ret);
+#endif
```

Comment:
These are causing a warning when included by a driver that already has a LOG declared for it

``
zephyr/include/zephyr/logging/log.h:438:17: warning: declaration of '__log_current_dynamic_data' shadows a global declaration [-Wshadow]
  438 |                 __log_current_dynamic_data __unused =                         \
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/zephyr/drivers/clock_management/clock_driver.h:308:9: note: in expansion of macro 'LOG_MODULE_DECLARE'
  308 |         LOG_MODULE_DECLARE(clock_management, CONFIG_CLOCK_MANAGEMENT_LOG_LEVEL);
      |         ^~~~~~~~~~~~~~~~~~
zephyr/include/zephyr/logging/log.h:438:17: note: shadowed declaration is here
  438 |                 __log_current_dynamic_data __unused =                         \
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/zephyr/logging/log.h:397:9: note: in expansion of macro 'LOG_MODULE_DECLARE'
  397 |         LOG_MODULE_DECLARE(__VA_ARGS__)
      |         ^~~~~~~~~~~~~~~~~~
spi_cdns.c:17:1: note: in expansion of macro 'LOG_MODULE_REGISTER'
   17 | LOG_MODULE_REGISTER(spi_cadence, CONFIG_SPI_LOG_LEVEL);
      | ^~~~~~~~~~~~~~~~~~~
```

Reply:
Reproduced this, good catch- it seems you can cause this by registering the log module before you include the clock management header:

```diff
diff --git a/drivers/serial/uart_mcux_flexcomm.c b/drivers/serial/uart_mcux_flexcomm.c
index 7d34bc774b6..f94eb072eab 100644
--- a/drivers/serial/uart_mcux_flexcomm.c
+++ b/drivers/serial/uart_mcux_flexcomm.c
@@ -19,6 +19,10 @@
 #include <soc.h>
 #include <fsl_device_registers.h>
 #include <zephyr/drivers/pinctrl.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_mcux_flexcomm, CONFIG_UART_LOG_LEVEL);
+
 #include <zephyr/drivers/clock_management.h>
```

We can fix this by silencing shadow warnings temporarily, like is done here: https://github.com/zephyrproject-rtos/zephyr/blob/b88d0bf8c7ee5fa308e9c6c8642d3efc6cb7064e/include/zephyr/debug/stack.h#L24

Pushed an update with this fix